test/protocol-tests/src/test/java/software/amazon/awssdk/protocol/tests/connection/SyncClientConnectionInterruptionTest.java

```diff
@@ -0,0 +1,253 @@
+                                                                     .dynamicPort());
+    @BeforeEach
+    public void setup() {
+        mockServer.start();
+        stubPostRequest(".*", aResponse(), "{}");
+    }
+
+    @Test
+    void connectionPoolsGetsReusedWhenInterruptedWith_1_MaxConnection() throws Exception {
+        Integer LONG_DELAY = 1500;
+
+        String urlRegex = "/2016-03-11/allTypes";
+        stubPostRequest(urlRegex, aResponse().withFixedDelay(LONG_DELAY), SAMPLE_BODY);
+        SdkHttpClient httpClient = ApacheHttpClient.builder().maxConnections(1).build();
+        ProtocolRestJsonClient client = getClient(httpClient, Duration.ofMillis(2L * LONG_DELAY)).build();
+
+        ExecutorService executorService = Executors.newFixedThreadPool(5);
+        Future<?> toBeInterruptedFuture = executorService.submit(() -> client.allTypes());
+        unInterruptedSleep(LONG_DELAY - LONG_DELAY / 5);
+        toBeInterruptedFuture.cancel(true);
+        // Make sure thread start the Http connections
+        unInterruptedSleep(50);
+        AllTypesResponse allTypesResponse = client.allTypes();
+        assertThat(allTypesResponse.stringMember()).isEqualTo("resultString");
+        executorService.shutdownNow();
+    }
+
+    @Test
+    void connectionPoolsGetsReusedWhenInterruptedWith_Multiple_MaxConnection() throws Exception {
+        Integer LONG_DELAY = 1000;
+        Integer VERY_VERY_LONG_DELAY = LONG_DELAY * 5;
+        stubPostRequest("/2016-03-11/allTypes", aResponse().withFixedDelay(LONG_DELAY), SAMPLE_BODY);
+        stubPostRequest("/2016-03-11/JsonValuesOperation", aResponse().withFixedDelay(VERY_VERY_LONG_DELAY), SAMPLE_BODY);
+        SdkHttpClient httpClient = ApacheHttpClient.builder().maxConnections(3).build();
+        Duration timeOutDuration = Duration.ofMillis(2L * LONG_DELAY);
+        ProtocolRestJsonClient client = getClient(httpClient, timeOutDuration).build();
+
+        ExecutorService executorService = Executors.newFixedThreadPool(5);
+        Future<?> toBeInterruptedFuture0 = executorService.submit(() -> client.allTypes());
+        Future<?> toBeInterruptedFuture1 = executorService.submit(() -> client.allTypes());
+        Future<?> toBeInterruptedFuture2 = executorService.submit(() -> client.allTypes());
+        unInterruptedSleep(50);
+        executorService.submit(() -> client.jsonValuesOperation());
+        unInterruptedSleep(LONG_DELAY / 2);
+        toBeInterruptedFuture0.cancel(true);
+        toBeInterruptedFuture1.cancel(true);
+        toBeInterruptedFuture2.cancel(true);
+        unInterruptedSleep(LONG_DELAY / 2);
+        // Make sure thread start the Http connections
+        AllTypesResponse allTypesResponse = client.allTypes();
+        assertThat(allTypesResponse.stringMember()).isEqualTo("resultString");
+        executorService.shutdownNow();
+    }
+
+    @Test
+    void interruptionWhenWaitingForLease_AbortsImmediately() throws InterruptedException {
```

Comment:
How do we know the request is waiting for a connection from the pool when it gets aborted?